File: src-tauri/src/db/update.rs
use crate::db::{
    schema,
    structs::{Event, Participant, Team},
    Database,
};
use crate::error::Error;
use crate::prelude::*;
use diesel::prelude::*;

impl Database {
    /// Update a team in the database.
    ///
    /// # Arguments
    ///
    /// * `id` - The ID of the team to update.
    /// * `name` - The new name of the team.
    /// * `points` - The new points of the team.
    ///
    /// # Returns
    ///
    /// The updated team.
    ///
    /// # Errors
    ///
    /// * If the team does not exist.
    /// * If the update fails.
    /// * If the connection to the database fails.
    pub fn update_team(&self, id: i32, name: String, points: i32) -> Result<Team> {
        use schema::team::dsl;

        let mut connection = self.connect()?;
        let updated_team = diesel::update(dsl::team.filter(dsl::id.eq(id)))
            .set((dsl::name.eq(name), dsl::points.eq(points)))
            .get_result(&mut connection)
            .map_err(|e| Error::DatabaseUpdateEntryFailure(e.to_string()))?;

        Ok(updated_team)
    }

    /// Update an event in the database.
    ///
    /// # Arguments
    ///
    /// * `id` - The ID of the event to update.
    /// * `name` - The new name of the event.
    /// * `event_type` - The new event type of the event.
    ///
    /// # Returns
    ///
    /// The updated event.
    ///
    /// # Errors
    ///
    /// * If the event does not exist.
    /// * If the update fails.
    /// * If the connection to the database fails.
    pub fn update_event(&self, id: i32, name: String, event_type: String) -> Result<Event> {
        use schema::event::dsl;

        let mut connection = self.connect()?;
        let updated_event = diesel::update(dsl::event.filter(dsl::id.eq(id)))
            .set((dsl::name.eq(name), dsl::event_type.eq(event_type)))
            .get_result(&mut connection)
            .map_err(|e| Error::DatabaseUpdateEntryFailure(e.to_string()))?;

        Ok(updated_event)
    }

    /// Update a participant in the database.
    ///
    /// # Arguments
    ///
    /// * `id` - The ID of the participant to update.
    /// * `team_id` - The new team ID of the participant.
    /// * `first_name` - The new first name of the participant.
    /// * `last_name` - The new last name of the participant.
    ///
    /// # Returns
    ///
    /// The updated participant.
    ///
    /// # Errors
    ///
    /// * If the participant does not exist.
    /// * If the update fails.
    /// * If the connection to the database fails.
    pub fn update_participant(
        &self,
        id: i32,
        team_id: i32,
        first_name: String,
        last_name: String,
    ) -> Result<Participant> {
        use schema::participant::dsl;

        let mut connection = self.connect()?;
        let updated_participant = diesel::update(dsl::participant.filter(dsl::id.eq(id)))
            .set((
                dsl::team_id.eq(team_id),
                dsl::first_name.eq(first_name),
                dsl::last_name.eq(last_name),
            ))
            .get_result(&mut connection)
            .map_err(|e| Error::DatabaseUpdateEntryFailure(e.to_string()))?;

        Ok(updated_participant)
    }
}

File: src/lib/database/update.ts
import { invoke as i } from '@tauri-apps/api/tauri';
import { handle_error, type Team, type Event, type Participant } from './';

export async function update_team({ id, name, points }: Team) {
	try {
		const db_team = await i<Team>('update_team', { id, name, points });

		alert(`Team ${db_team.name} updated successfully.`);
		return db_team;
	} catch (e) {
		handle_error(e as Record<string, string>);
	}
}

export async function update_participant({ id, team_id, first_name, last_name }: Participant) {
	try {
		const db_participant = await i<Participant>('update_participant', {
			id,
			team_id,
			first_name,
			last_name
		});
		alert(
			`Participant ${db_participant.first_name} ${db_participant.last_name} updated successfully.`
		);
		return db_participant;
	} catch (e) {
		handle_error(e as Record<string, string>);
	}
}

export async function update_event({ id, name, event_type }: Event) {
	try {
		const db_event = await i<Event>('update_event', { id, name, event_type });

		alert(`Event ${db_event.name} updated successfully.`);
		return db_event;
	} catch (e) {
		handle_error(e as Record<string, string>);
	}
}
